Refactored router tests

nibra opened the pull request:
- less assertions per test
- speaking names for most test methods
- test JRouter instead of JUri
- use Reflection instead of mocks to test internals

nibra committed:
Refactored router tests
  - less assertions per test
  - speaking names for most test methods
  - test JRouter instead of JUri
  - use Reflection instead of mocks to test internals

wilsonge commented on tests/unit/suites/libraries/cms/router/JRouterAdministratorTest.php:
```diff
@@ -73,41 +74,30 @@ protected function tearDown()
-			'JRouterAdministrator::parse() returns an empty array.'
-		);
+		$vars = $this->object->parse($uri);
+		$this->assertTrue(is_array($vars));
```
Does using ```assertInstanceOf``` work here. Personally I'd prefer that to this

mbabker replied:
No.  There is something like assertInternalType or something that works
with scalar objects.

On Friday, July 17, 2015, George Wilson <notifications@github.com> wrote:

> In tests/unit/suites/libraries/cms/router/JRouterAdministratorTest.php
> <https://github.com/joomla/joomla-cms/pull/7455#discussion_r34884755>:
>
> >  	 * @since   3.0
> >  	 */
> >  	public function testParse()
> >  	{
> >  		$uri = JUri::getInstance('http://localhost');
> >
> > -		$this->assertThat(
> > -			$this->object->parse($uri),
> > -			$this->isType('array'),
> > -			'JRouterAdministrator::parse() returns an empty array.'
> > -		);
> > +		$vars = $this->object->parse($uri);
> > +		$this->assertTrue(is_array($vars));
>
> Does using assertInstanceOf work here. Personally I'd prefer that to this
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/7455/files#r34884755>.
>

wilsonge replied:
```assertInternalType``` only works for variables inside a class. It's no good for the result of a function. I guess it's this or the ```$this->assertThat``` combined with ```$this->isType``` that we had before

mbabker replied:
No, that one isn't for class member vars; those are usually
assertAttributeWhatever assertions.  That one should work just fine for
this case.

On Fri, Jul 17, 2015 at 10:57 AM, George Wilson <notifications@github.com>
wrote:

> In tests/unit/suites/libraries/cms/router/JRouterAdministratorTest.php
> <https://github.com/joomla/joomla-cms/pull/7455#discussion_r34896961>:
>
> >  	 * @since   3.0
> >  	 */
> >  	public function testParse()
> >  	{
> >  		$uri = JUri::getInstance('http://localhost');
> >
> > -		$this->assertThat(
> > -			$this->object->parse($uri),
> > -			$this->isType('array'),
> > -			'JRouterAdministrator::parse() returns an empty array.'
> > -		);
> > +		$vars = $this->object->parse($uri);
> > +		$this->assertTrue(is_array($vars));
>
> assertInternalType only works for variables inside a class. It's no good
> for the result of a function. I guess it's this or the $this->assertThat
> combined with $this->isType that we had before
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/7455/files#r34896961>.
>

nibra replied:
Right. I had to rework the test anyway because of the misleading message. Usually, I prefer the most specific assertion, but in this case, I find `assertTrue` easiest to read.

nibra replied:
However, I can change it back, if you want me to.

nibra replied:
George, you're talking about `assertAttributeInternalType`, I assume. Going to rewrite the test using `assertInternalType`.

mbabker replied:
I'm impartial either way, that assertInternalType method reads kinda funny to me sometimes.  assertTrue(is_array()) at least makes it clear what you're testing without some of the fancier assertions.

nibra replied:
I understand `assertInternalType` to assert the (internal) type *without implicit cast*.

nibra replied:
So, should I leave it or change it? I'm happy with both.

wilsonge replied:
Nah leave it as it is - it's not worth the effort for something so small in the bigger picture.

wilsonge commented on tests/unit/suites/libraries/cms/router/JRouterSiteTest.php:
```diff
@@ -7,24 +7,26 @@
+	 * @since  4.0
 	 */
-	protected $object;
+	private $input;
```
I don't think you are using this class var?

nibra replied:
Correct - removing it.
Usually, PhpStorm marks unused private properties, but it didn't do it here, so I overlooked it.

wilsonge commented on tests/unit/suites/libraries/cms/router/JRouterTest.php:
```diff
@@ -30,104 +29,131 @@ class JRouterTest extends TestCase
+		 * The following is needed to work around a bug in JApplicationWeb::detectRequestUri()
+		 * @see https://github.com/joomla-projects/joomla-pythagoras/issues/2
+		 */
+		if (!isset($_SERVER['HTTP_HOST']))
```
If we're setting this into server we really ought to be tearing it back down again - even with just an empty string we risk influencing future tests

nibra replied:
Fixing it.

wilsonge commented on tests/unit/suites/libraries/cms/router/JRouterTest.php:
```diff
@@ -30,104 +29,131 @@ class JRouterTest extends TestCase
+	 */
+	public function testProperTypeAndMode($client)
+	{
+		$this->markTestSkipped('Untestable due to global instance cache not clearable.');
```
What does this comment mean? Are you referring to the static here https://github.com/joomla/joomla-cms/blob/staging/libraries/cms/router/router.php#L136 if so can't we use Reflection to reset it? http://php.net/manual/en/reflectionclass.setstaticpropertyvalue.php

nibra replied:
Yes, that seems to work.I just didn't think about using Reflection on non-objects.

wilsonge commented:
Couple of small comments - but overall looks really awesome :) ty!

nibra committed:
Minor fixes
 - removed unused property
 - avoided side effect with global var
 - used reflection to clear instance cache

nibra commented:
Fixed the issues, you found, George.

wilsonge commented:
Awesome. I've pinged @mbabker just to make sure I haven't missed anything. And will merge as soon as he gives the goahead :)

mbabker commented:
Don't wait for me, I'm conferencing this weekend and still owe another team a review/merge on production code.
